qemu/tests/in_place_upgrade.py

```diff
@@ -90,8 +94,25 @@ def run(test, params, env):
         upgrade_test.post_upgrade_restore(test)
         # post checking
         upgrade_test.post_upgrade_check(test, post_release)
+        post_rhel_ver = upgrade_test.run_guest_cmd(check_rhel_ver)
         vm.verify_kernel_crash()
     finally:
-        if upgrade_test.session:
-            upgrade_test.session.close()
-        vm.destroy()
+        vm.graceful_shutdown(timeout=300)
+        try:
+            rt_dir = data_dir.get_data_dir() + "/"
+            image_format = params.get("image_format")
+            old_name = params.get("image_name")
+            old_name = "%s.%s" % (old_name, image_format)
+            upgraded_name = old_name.replace(pre_rhel_ver, post_rhel_ver)
+            process.run(params.get("image_clone_command") %
+                        (rt_dir + old_name, rt_dir + upgraded_name))
+            if "ovmf" in params.get("id"):
+                ovmf_file_old = rt_dir + params.get("main_vm") + "_"
+                ovmf_file_old += params.get("image_name")[7:] + "_"
+                ovmf_file_old += params.get("image_format")
+                ovmf_file_old += "_filesystem_VARS.fd"
+                ovmf_file = ovmf_file_old.replace(pre_rhel_ver, post_rhel_ver)
+                process.run(params.get("image_clone_command") %
+                            (ovmf_file_old, ovmf_file))
```

Comment:
Not sure we need to handle this. firmware file will automatically be generated if it doesn't exist, and the format depends on the filesystem.

Reply:
Yes, will remove it. Thanks.